luigi/contrib/batch.py

```diff
@@ -186,28 +192,121 @@ def register_job_definition(self, json_fpath):
+
+    :param str job_name: name of specific job, for tracking in the queue and logs.
+
+    :param int poll_time: interval in seconds to check a status of jobs.
+
+    :param str queue: name (arn) of the job queue.
+
+        .. code:: python
+
+            queue = 'arn:aws:batch:<region>:<user_id>:job-queue/<job_queue_name>'
+
+    :param str array_properties: array properties for the job in JSON format.
+
+        .. code:: python
+
+            array_properties = json.dumps(
+                {
+                    'size': 123
+                }
+            )
+
+    :param str depends_on: dependencies for the job in JSON format.
+
+        .. code:: python
+
+            depends_on = json.dumps(
+                [
+                    {
+                        'jobId': 'string',
+                        'type': 'N_TO_N'
+                    },
+                    {
+                        'jobId': 'string',
+                        'type': 'SEQUENTIAL'
+                    }
+                ]
+            )
+
+    :param str container_overrides: container overrides for the job in JSON format.
+
+        .. code:: python
+
+            container_overrides = json.dumps(
+                {
+                    'vcpus': 123,
+                    'memory': 123,
+                    'command': [
+                        'string'
+                    ],
+                    'environment': [
+                        {
+                            'name': 'string',
+                            'value': 'string'
+                        }
+                    ]
+                }
+            )
+
+    :param str retry_strategy: retry strategy for the job in JSON format.
+
+        .. code:: python
+
+            retry_strategy = json.dumps(
+                {
+                    'attempts': 123
+                }
+            )
+
+    :param str timeout: timeout configuration for the job in JSON format.
+
+        .. code:: python
+
+            timeout = json.dumps(
+                {
+                    'attemptDurationSeconds': 123
+                }
+            )
 
     """
+
     job_definition = luigi.Parameter()
     job_name = luigi.OptionalParameter(default=None)
     poll_time = luigi.IntParameter(default=POLL_TIME)
+    queue = luigi.OptionalParameter(default=None)
+    array_properties = luigi.OptionalParameter(default='{}')
+    depends_on = luigi.OptionalParameter(default='[]')
+    container_overrides = luigi.OptionalParameter(default='{}')
+    retry_strategy = luigi.OptionalParameter(default='{}')
+    timeout = luigi.OptionalParameter(default='{}')
 
     def run(self):
         bc = BatchClient(self.poll_time)
         job_id = bc.submit_job(
             self.job_definition,
             self.parameters,
-            job_name=self.job_name)
+            job_name=self.job_name,
+            queue=self.queue,
+            array_properties=json.loads(self.array_properties),
```

Comment:
You say there is an issue here when using `DictParameter` due to type checking?

I've encountered something similar. I wonder what we can do to DictParameter to resolve the returned object type. Would you happen to have any suggestions?

It feels odd to be using `OptionalParameter` here (although functional). That said, I don't think this is a blocker. But perhaps at least worth an in-line comment for reason.

Reply:
@dlstadther Thanks for your comments!

> You say there is an issue here when using DictParameter due to type checking?

Yes, boto3 raises validation error when using DictParameter.
```
ParamValidationError: Parameter validation failed:
Invalid type for parameter containerOverrides, value: <FrozenOrderedDict OrderedDict([('vcpus', 123)])>, type: <class 'luigi.parameter._FrozenOrderedDict'>, valid types: <class 'dict'>
```

> It feels odd to be using OptionalParameter here (although functional). That said, I don't think this is a blocker. But perhaps at least worth an in-line comment for reason.

I think so too.
Since aws api frequently uses json format, I want to fix it conveniently if possible.

> I've encountered something similar. I wonder what we can do to DictParameter to resolve the returned object type. Would you happen to have any suggestions?

How about JSON compatible parameter?
1. User can set JSON object such as dictionary, list, string, numerical number, and boolean.
2. The parameter hold's value as json serialized string.
3. In the task class, parameters can be accessed as deserialized objects.

```
class FooTask(luigi.Task):
    param1 = luigi.JSONParameter()

    def run(self):
        print(self.param1)


foo = FooTask(param1={'key': 'value'})
foo.run()  # {'key': 'value'}
```

Reply:
In that example, the only usage difference between JsonParameter and DictParameter is Dict's ability to allow an element to be a custom object?

There's definitely a tight relationship between the two.

I'd like to involve some other Luigi-contributors around this:
@Tarrasch 
@joemeszaros 
@nugend

Reply:
> In that example, the only usage difference between JsonParameter and DictParameter is Dict's ability to allow an element to be a custom object?

Yes, indeed.

If you want to develop based on DictParameter, I want a function to restore from _FrozenOrderedDict to pure python dictionary.
For example _recursively_defreeze(value).

Reply:
Why is boto3 restricting to only values of type dict? Shouldn't it be looking for Mapping or MutableMapping from collections (depending on if it actually needs to modify or not).

Reply:
Is it possible that's an upstream issue then?

Reply:
I opened new issue about DictParameter #2562 .
And I added in-line comments.